tests/PrefixCacheBootstrapperTest.php

```diff
@@ -0,0 +1,345 @@
+<?php
+
+declare(strict_types=1);
+
+use Illuminate\Cache\CacheManager;
+use Illuminate\Support\Facades\Event;
+use Stancl\Tenancy\Events\TenancyEnded;
+use Stancl\Tenancy\Tests\Etc\CacheService;
+use Stancl\Tenancy\Events\TenancyInitialized;
+use Stancl\Tenancy\Listeners\BootstrapTenancy;
+use Stancl\Tenancy\Listeners\RevertToCentralContext;
+use Stancl\Tenancy\CacheManager as TenancyCacheManager;
+use Stancl\Tenancy\Tests\Etc\SpecificCacheStoreService;
+use Stancl\Tenancy\Bootstrappers\PrefixCacheTenancyBootstrapper;
+
+beforeEach(function () {
+    config([
+        'tenancy.bootstrappers' => [
+            PrefixCacheTenancyBootstrapper::class
+        ],
+        'cache.default' => $cacheDriver = 'redis',
+    ]);
+
+    PrefixCacheTenancyBootstrapper::$tenantCacheStores = [$cacheDriver];
+    PrefixCacheTenancyBootstrapper::$prefixGenerator = null;
+
+    TenancyCacheManager::$addTags = false;
+
+    Event::listen(TenancyInitialized::class, BootstrapTenancy::class);
+    Event::listen(TenancyEnded::class, RevertToCentralContext::class);
+});
+
+afterEach(function () {
+    PrefixCacheTenancyBootstrapper::$tenantCacheStores = [];
+    PrefixCacheTenancyBootstrapper::$prefixGenerator = null;
+
+    TenancyCacheManager::$addTags = false;
+});
+
+test('Tenancy overrides CacheManager', function() {
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->initialize(Tenant::create(['id' => 'first']));
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->initialize(Tenant::create(['id' => 'second']));
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
+
+    tenancy()->end();
+
+    expect(app('cache')::class)->toBe(TenancyCacheManager::class);
+    expect(app(CacheManager::class)::class)->toBe(TenancyCacheManager::class);
```

Comment:
Is the point of these repeated assertions ensuring that the cache manager is always overridden, no matter what happens?

And on a related note: is there an easy way for people to disable the cache manager override (in case they want to use a custom CacheManager)? Should be a static property/config key probably. Maybe putting it inside https://github.com/archtechx/tenancy/blob/6a26f712e797aa9ed79c54be281c4fb1069328af/assets/config.php#L191 would work well.

Reply:
> Is the point of these repeated assertions ensuring that the cache manager is always overridden, no matter what happens?

Yes

> is there an easy way for people to disable the cache manager override (in case they want to use a custom CacheManager)?

No. The config key idea seems good, though we'll need to mention that CacheTagsBootstrapper only works with our custom CacheManager. Or we can make the CacheManager implement some interface that requires the manager to have the static `$addTags` property (e.g. UsesTags) and require the CacheManager that implements that interface in the bootstrapper.

Reply:
> though we'll need to mention that CacheTagsBootstrapper only works with our custom CacheManager

That's fine.

> Or we can make the CacheManager implement some interface that requires the manager to have the static $addTags property (e.g. UsesTags)

Interfaces can't have properties, only methods. And sounds unnecessarily complex. Leaving a note that X feature only works with that config enabled is completely fine. 99% of people won't ever disable it, it's just to give people who need more customization the option to do so.

Reply:
The config key is there and it's used now, but I didn't add any tests for it yet – I'll finish this tomorrow

Reply:
Added a todo for the test https://github.com/archtechx/tenancy/pull/1014/commits/effb6a81a1899aa54aa844be8be2ade4d3e1b1bf